brushfire-core/src/main/scala/com/stripe/brushfire/Tree.scala

```diff
@@ -31,6 +31,82 @@ case class Tree[K, V, T](root: Node[K, V, T]) {
     }
   }
 
+  /**
+   * Prune a tree to minimize validation error.
+   *
+   * Recursively replaces each split with a leaf when it would have a lower error than the sum of the child leaves
+   * errors.
+   *
+   * @param validationData A Map from leaf index to validation data.
+   * @param voter to create predictions from target distributions.
+   * @param error to calculate an error statistic given observations (validation) and predictions (training).
+   * @return The new, pruned tree.
+   */
+  def prune[P, E](validationData: Map[Int, T], voter: Voter[T, P], error: Error[T, P, E])(implicit targetMonoid: Monoid[T], errorOrdering: Ordering[E]): Tree[K, V, T] = {
+    Tree(pruneNode(validationData, this.root, voter, error)).renumberLeaves
+  }
+
+  /**
+   * Prune a tree to minimize validation error, starting from given root node.
+   *
+   * This method recursively traverses the tree from the root, branching on splits, until it finds leaves, then goes back
+   * down the tree combining leaves when such a combination would reduce validation error.
+   *
+   * @param validationData Map from leaf index to validation data.
+   * @param start The root node of the tree.
+   * @return A node at the root of the new, pruned tree.
+   */
+  def pruneNode[P, E](validationData: Map[Int, T], start: Node[K, V, T], voter: Voter[T, P], error: Error[T, P, E])(implicit targetMonoid: Monoid[T], errorOrdering: Ordering[E]): Node[K, V, T] = {
+    var vData = validationData
+    start match {
+      case leaf: LeafNode[K, V, T] => leaf // Bounce at the bottom and start back up the tree.
+      case SplitNode(children) => {
+        val newNode = SplitNode(children.map { case (k, p, n) => (k, p, pruneNode(vData, n, voter, error)) })
+        val childLeaves = newNode.children.collect { case (k, v, s: LeafNode[K, V, T]) => (k, v, s) }
+        if (childLeaves.size == newNode.children.size)
+          pruneLevel(newNode, childLeaves, vData, voter, error) match { case (v, n) => vData = v; n }
```

Comment:
so, if I understand this right, the vData being returned from child N is being fed into child N+1? Rather than using a local `var` for this I would prefer to see a `foldLeft` which makes that pattern more explicit.

(Sorry for all the style nits).